### PR TITLE
Add AffineNormalizer for action. Clamp heightmap.

### PR DESCRIPTION
**Pull request by hongkai-dai**
This change is [Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@blake-wulfe-tri for review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-:-NS8Mik77dnhwfb9VWrP:b9e17as)* status: 0 of 3 files reviewed, all discussions resolved (waiting on @blake-wulfe-tri)

**Review by blake-wulfe-tri — CHANGES_REQUESTED** [review]
:lgtm_strong: 

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-:-NS8yVkrBnhOvga2eYLk:b-uwkowd)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @hongkai-dai)

___
*[`mani_skill2/dough_model_learning/train.py` line 70 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NS8x4HF6zBSYqTxnMAS:-NS8x4HF6zBSYqTxnMAT:bh8zewm) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/115588672429c4f98ef86939526e5222ab6be056/mani_skill2/dough_model_learning/train.py#L70)):*
> ```Python
> 
> 
> def get_action_normalizer(train_loader: torch.utils.data.Dataset):
> ```

this should probably compute the normalizer based on the full dataset right? you could do that with something like:
```python
actions = torch.cat([batch["actions"] for batch in loader])
return AffineNormalizer.from_data(actions)
```
it might take a while to compute, but I think it's probably worthwhile
Another issue might be that it doesn't all fit in memory, but since we should just be keeping the actions in memory I don't think it'll be a problem
___
*[`mani_skill2/dynamics/normalizers.py` line 70 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NS8xaK6B_np6JQ34D_N:-NS8xaK6B_np6JQ34D_O:b-9y3gdk) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/115588672429c4f98ef86939526e5222ab6be056/mani_skill2/dynamics/normalizers.py#L70)):*
> ```Python
>             minimum along dim=0 as lo and up
>         """
>         return cls(torch.min(data, dim=0).values, torch.max(data, dim=0).values)
> ```

btw a lot of the time there are outliers that result in squashing the rest of the data to a small range and making it hard to learn. What do you think about taking the min/max as the 1% and 99% percentile instead, and clipping the data to that range? Might not be a problem yet, but I suspect it may be

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dough_model_learning/train.py` line 70 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NS8x4HF6zBSYqTxnMAS:-NS9J3hp4WsQui_DnYZS:b-yh16l6) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/115588672429c4f98ef86939526e5222ab6be056/mani_skill2/dough_model_learning/train.py#L70)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

this should probably compute the normalizer based on the full dataset right? you could do that with something like:
```python
actions = torch.cat([batch["actions"] for batch in loader])
return AffineNormalizer.from_data(actions)
```
it might take a while to compute, but I think it's probably worthwhile
Another issue might be that it doesn't all fit in memory, but since we should just be keeping the actions in memory I don't think it'll be a problem
</blockquote></details>

Done. Good call.
___
*[`mani_skill2/dynamics/normalizers.py` line 70 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NS8xaK6B_np6JQ34D_N:-NS9_vxkE_4FGf_YpJ3m:b-ohdo8) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/115588672429c4f98ef86939526e5222ab6be056/mani_skill2/dynamics/normalizers.py#L70)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw a lot of the time there are outliers that result in squashing the rest of the data to a small range and making it hard to learn. What do you think about taking the min/max as the 1% and 99% percentile instead, and clipping the data to that range? Might not be a problem yet, but I suspect it may be
</blockquote></details>

Done, good call.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @blake-wulfe-tri)

*[a discussion](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSDxTSh6T6nMjD1TbRa:-NSDxTSh6T6nMjD1TbRb:b-umrk4s) (no related file):*
Do not merge this yet, as we have problem to load state dict for the AffineNormalizer.

___

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32)* status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @blake-wulfe-tri)

*[a discussion](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSDxTSh6T6nMjD1TbRa:-NSHPwa-72SZvG_tyUhB:b-psyi2b) (no related file):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Do not merge this yet, as we have problem to load state dict for the AffineNormalizer.
</blockquote></details>

This PR is ready for review again.

___

**Review by blake-wulfe-tri — CHANGES_REQUESTED** [review]
:lgtm: 

> Reviewed 2 of 7 files at r2, 6 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-:-NSHruC5BTmir5LucLf0:bg7igkv)* status: 8 of 10 files reviewed, 4 unresolved discussions (waiting on @hongkai-dai)

___
*[`mani_skill2/dough_model_learning/train.py` line 75 at r3](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSHqtlDAGTvRepx6zmI:-NSHqtlDAGTvRepx6zmJ:b5r6fjc) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/dbbc0e0b32be0827433267654aa47b640f7fa59e/mani_skill2/dough_model_learning/train.py#L75)):*
> ```Python
>     actions = torch.cat(
>         [batch["actions"] for batch in train_loader]
>         + [batch["actions"] for batch in val_loader]
> ```

we shouldn't use the validation loader for computing the stats b/c it invalidates the performance of the validation set as a useful measure (to some extent)
___
*[`mani_skill2/dynamics/normalizers.py` line 3 at r3](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSHrIe66EZSF4VXpKpP:-NSHrIe66EZSF4VXpKpQ:byp1uy2) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/dbbc0e0b32be0827433267654aa47b640f7fa59e/mani_skill2/dynamics/normalizers.py#L3)):*
> ```Python
> import abc
> import math
> import numpy as np
> ```

btw the built-in imports should go first, then a separation line for the installed packages (based on google style guide)

so like

```python
import abc
import math
import typing

import numpy as np
import torch
...
```
___
*[`mani_skill2/dynamics/normalizers.py` line 89 at r3](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSHrjXt3IrxWR-GSWW-:-NSHrjXt3IrxWR-GSWW0:bfl2reo) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/dbbc0e0b32be0827433267654aa47b640f7fa59e/mani_skill2/dynamics/normalizers.py#L89)):*
> ```Python
>         """
>         assert percentile >= 0
>         # Sort in ascending order.
> ```

btw maybe use quantile instead? https://pytorch.org/docs/stable/generated/torch.quantile.html

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32)* status: 8 of 10 files reviewed, 3 unresolved discussions (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dough_model_learning/train.py` line 75 at r3](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSHqtlDAGTvRepx6zmI:-NSJY9OrBW74TyW_K3Ld:b-896fix) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/dbbc0e0b32be0827433267654aa47b640f7fa59e/mani_skill2/dough_model_learning/train.py#L75)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

we shouldn't use the validation loader for computing the stats b/c it invalidates the performance of the validation set as a useful measure (to some extent)
</blockquote></details>

Done.
___
*[`mani_skill2/dynamics/normalizers.py` line 89 at r3](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSHrjXt3IrxWR-GSWW-:-NSJYQ6z1c51zFveDyq9:b-5b1wkx) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/dbbc0e0b32be0827433267654aa47b640f7fa59e/mani_skill2/dynamics/normalizers.py#L89)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw maybe use quantile instead? https://pytorch.org/docs/stable/generated/torch.quantile.html
</blockquote></details>

I think the motivation here is to exclude the outlier. So if we have some data that every entry (except one outlier) is within 0 to 1, but the outlier is 100, then use quantile(0.05) will give us 95 as the upper bound, but using this sorting and excluding the top 5% entries will still give us 1 as the upper bound, hence it is more robust to the outlier

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32)* status: 6 of 10 files reviewed, 2 unresolved discussions (waiting on @hongkai-dai)

___
*[`mani_skill2/dynamics/normalizers.py` line 89 at r3](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/32#-NSHrjXt3IrxWR-GSWW-:-NSKFy2F6P6p0uu7qZ6r:bfq9ba) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/dbbc0e0b32be0827433267654aa47b640f7fa59e/mani_skill2/dynamics/normalizers.py#L89)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think the motivation here is to exclude the outlier. So if we have some data that every entry (except one outlier) is within 0 to 1, but the outlier is 100, then use quantile(0.05) will give us 95 as the upper bound, but using this sorting and excluding the top 5% entries will still give us 1 as the upper bound, hence it is more robust to the outlier
</blockquote></details>

sounds good.
